### PR TITLE
Fingerprint: config parameter for interface and native method for IP address

### PR DESCRIPTION
**Pull request by apognu**
This PR tries to respond to hashicorp/nomad#170 and hashicorp/nomad#158.

I added a configuration directive to the client conf file to set the network interface used for fingerprinting.

```
client {
  iface = "enp2s0"
}
```

It also adds a second method for retrieving the IP address of this network card (before we try ifconfig) using net.Interfaces(). This is not perfect yet since I totally ignore IPv6 addresses, but it's a base to develop from.

### CODE REVIEW

File: client/fingerprint/network_unix.go [hunk 1]
```diff
@@ -129,6 +133,54 @@ func (f *NetworkFingerprint) linkSpeedEthtool(path, device string) int {
 	return mbs
 }
 
+// ipAddress returns the first IPv4 address on the configured default interface
+// Tries Golang native functions and falls back onto ifconfig
+func (f *NetworkFingerprint) ipAddress(device string) string {
+	if ip, err := f.nativeIpAddress(device); err == nil {
```
Comment (achanda) [inline client/fingerprint/network_unix.go]: shouldn't that be `err != nil`?

Comment (apognu) [inline client/fingerprint/network_unix.go]: Well, no, that function should return the IP if there was no error, so if err == nil. Look at the if body.

If err is different than nil, its execution would continue, and we would try the ifconfig method.

Comment (achanda) [inline client/fingerprint/network_unix.go]: ah, I see. Thanks.

File: client/fingerprint/network_unix.go [hunk 2]
```diff
@@ -38,10 +39,13 @@ func (f *NetworkFingerprint) Fingerprint(cfg *config.Config, node *structs.Node)
 	if "darwin" == runtime.GOOS {
 		defaultDevice = "en0"
 	}
+	if cfg.Iface != "" {
```
Comment (dadgar) [inline client/fingerprint/network_unix.go]: Put a comment above to the tune of:

// Allow the user to override the default network device.

File: client/fingerprint/network_unix.go [hunk 3]
```diff
@@ -129,6 +133,54 @@ func (f *NetworkFingerprint) linkSpeedEthtool(path, device string) int {
...
+	// Find IP address on configured interface
+	var ip string
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return "", errors.New("could not retrieve interface list")
+	}
+
+	// TODO: should we handle IPv6 here? How do we determine precedence?
+	for _, i := range ifaces {
+		if i.Name == device {
```
Comment (dadgar) [inline client/fingerprint/network_unix.go]: if i.Name != device { continue }

Lets us un-indent that block

File: client/config/config.go [hunk 4]
```diff
@@ -31,6 +31,9 @@ type Config struct {
 	// Region is the clients region
 	Region string
 
+	// Network interface to be used in network fingerprinting
+	Iface string
```
Comment (dadgar) [inline client/config/config.go]: Lets change Iface to "NetworkInterface" for long term clarity

**Pull request merged**